plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,402 @@
+static unsigned int smp_numcores;
+void lcpu_start(struct arm_cpu *cpu);
+void __noreturn lcpu_entry_default(void);
+
+static __lcpuid ukplat_lcpu_idx(void);
+
+static int arm64_lcpu_ipi_run_handler(void *args __unused, struct __regs *regs)
+{
+	struct ukplat_lcpu_func *i;
+	struct uk_list_head *fnlist;
+
+	__lcpuid this_cpu_idx = ukplat_lcpu_idx();
+
+	cpus[ukplat_lcpu_id()].state = LCPU_STATE_BUSY;
+	ukarch_spin_unlock(&cpus[this_cpu_idx].lock);
+
+	fnlist = &cpus[this_cpu_idx].fnlist;
+
+	uk_list_for_each_entry(i, fnlist, lentry) {
+		i->fn(regs, i);
+		if (i == uk_list_first_entry(fnlist,
+				struct ukplat_lcpu_func, lentry))
+			break;
+	}
+
+	return 1;
+}
+
+int smp_init(void *dtb)
+{
+	int fdt_cpu;
+	const fdt32_t *naddr_prop, *nsize_prop, *id_reg;
+	const struct fdt_property *cpu_prop;
+	int naddr, nsize = 0, cell;
+	__lcpuid cpu_idx;
+	int subnode;
+	char bsp_found = 0;
+
+	/* Init the info for the bootstraping cpu */
+	bsp_cpu_id = ukplat_lcpu_id();
+	cpus[0].id = bsp_cpu_id;
+	cpus[0].state = LCPU_STATE_BUSY;
+	ukarch_spin_init(&cpus[0].lock);
+	uk_pr_info("Bootstrapping processor has the ID %d\n", bsp_cpu_id);
+
+	/* Init the secondary cpus vector */
+	for (cpu_idx = 1; cpu_idx < CONFIG_UKPLAT_LCPU_MAXCOUNT; cpu_idx++) {
+		cpus[cpu_idx].id = (__u64) -1;
+		cpus[cpu_idx].state = LCPU_STATE_OFFLINE;
+	}
+
+	/* Search for cpus in DTB */
+	fdt_cpu = fdt_path_offset(dtb, "/cpus");
+	if (fdt_cpu < 0) {
+		uk_pr_err("cpus node is not found in device tree\n");
+		return -1;
+	}
+
+	/* Get address cells */
+	naddr_prop = fdt_getprop(dtb,
+			fdt_cpu, "#address-cells", &cell);
+	if (naddr_prop == NULL) {
+		naddr = FDT_ADDR_CELLS_DEFAULT;
+		uk_pr_warn("Using default value 2 for cpu address-cells.\n");
+	} else {
+		naddr = fdt32_to_cpu(naddr_prop[0]);
+	}
+	if (naddr < 0 || naddr >= FDT_MAX_NCELLS) {
+		uk_pr_err("Invalid address-cells size.\n");
+		return -1;
+	}
+	if (naddr > FDT_ADDR_CELLS_DEFAULT) {
+		uk_pr_err("Address-cells size greater than 2 not supported yet.\n");
+		return -1;
+	}
+
+	/* Get size cells */
+	nsize_prop = fdt_getprop(dtb, fdt_cpu,
+			"#size-cells", &cell);
+	if (nsize_prop == NULL) {
+		nsize = FDT_SIZE_CELLS_DEFAULT;
+		uk_pr_warn("Using default value of 0 for cpu size-cells.\n");
+	} else {
+		nsize = fdt32_to_cpu(nsize_prop[0]);
+	}
+	/**
+	 *  According to the device-tree v0.3 specification, for the `size-cells`
+	 *  property of the /cpus node: "Value shall be 0. Specifies that no size is
+	 *  required in the reg property in children of this node."
+	 */
+	if (nsize != 0) {
+		uk_pr_err("Invalid size-cells value for cpu node.\n");
+		return -1;
+	}
+
+	/* Search all the cpu nodes in DTB */
+	cpu_idx = 0;
+	fdt_for_each_subnode(subnode, dtb, fdt_cpu) {
+		int prop_len = 0;
+		__u64 cpu_id;
+
+		cpu_prop = fdt_get_property(dtb, subnode,
+			"device_type", &prop_len);
+		if (!cpu_prop)
+			continue;
+		if (strcmp(cpu_prop->data, "cpu"))
+			continue;
+
+		cpu_prop = fdt_get_property(dtb, subnode,
```

Comment:
There is a case where one might have configured Unikraft for SMP but execute it on a single core system, eg:
```
qemu-system-aarch64 -machine virt,mte=on -cpu max -m 8M -monitor none -serial stdio -nographic -kernel build/app-helloworld_kvm-arm64
```
This currently breaks, because QEMU virt does not specify an `enable-method` node in this case:
```
        cpus {    
                #size-cells = <0x00>;    
                #address-cells = <0x01>;    
    
                cpu@0 {    
                        reg = <0x00>;    
                        compatible = "arm,cortex-a57";    
                        device_type = "cpu";    
                };    
        };
```
I am not sure if this is valid by the device-tree spec. If it is, or if we want to be forgiving, perhaps it would be worth adding a check where if there is a single cpu node we skip the check for `enable-method`. I don't have a strong opinion though, what do you think?

Reply:
The spec says about `enable-method` that: `This property is required for CPUs with a status property with a value of "disabled".` I think QEMU assumes that, in a single core system, the core can't be in the disabled state, so the `enable-method` field is not necessary. I will add a check so that the missing `enable-method` is ignored for the bsp.

Reply:
> The spec says about enable-method that: This property is required for CPUs with a status property with a value of "disabled".

Yes I am aware of that. What I meant that it is not clear to me by the spec whether the `enable-method` is mandatory for single-core systems.

>  I think QEMU assumes that, in a single core system, the core can't be in the disabled state, so the enable-method field is not necessary.

Yes, I think so too. The question is how lenient we want to be if we think that a platform breaks the spec, or if we think the spec is unclear.

Reply:
I took a second look and, in this scenario, my view would be to still check for the `enable-method` in the bootstrapping core and, if it is missing, throw a different message than we do for APs.
Right now, for secondary cores, if the enable method is missing we print "enable-method for core 0x%lx not set, core will not be enabled.". For the bsp, we should still make the check. However, if the `enable-method` is missing, the core is still enabled, so I would replace the error message with a warning one and simply print  "enable-method for core 0x%lx not set".

Reply:
Also on this topic, do you think it would be useful to print a warning if the user configured `UKPLAT_LCPU_MAXCOUNT` to a value greater than 1, but is trying to run on a single core system?

Reply:
> Also on this topic, do you think it would be useful to print a warning if the user configured `UKPLAT_LCPU_MAXCOUNT` to a value greater than 1, but is trying to run on a single core system?

No, that is not necessary. We can assume it to be the regular case that `MAXCOUNT` is greater than the systems number of cores.

BTW: Please wait with changes in the core API functions `ukplat_lcpu_*`. I have created a generic version of these functions based on your implementation that I will provide you in a few hours.

Reply:
Yes, sure. I just want to avoid that you are working on code that might be replaced. But everything strictly ARM related I do not touch. I am almost ready with the commit history. We can then decide how to proceed. Perhaps I do a separate PR for the generic parts and you change this PR to plug into that, if you are fine with that approach.